### ISSUE TITLE
Add --safe-except-for-sized-types?

### ISSUE THREAD

**Issue by nad** [agda/agda#4908]
Given that we have some long-standing consistency bugs related to sized types it makes sense to disallow sized types when `--safe` is active. However, I have quite a bit of code that uses sized types, and I'd like Agda to check that I'm not using other unsafe features. Thus it might make sense to add a flag like `--safe-except-for-sized-types`.

An alternative would be to add a more general flag with support for other unsafe features: `--safe-except-for=…`. Does anyone see a use for the more general variant?

A different approach would be to list the features that are allowed: `--language=…`. Then one could enumerate features like "user-defined data types", "pattern matching", and so on. However, implementing this option might be non-trivial.

**Comment by turion** [comment]
`--safe-except-for=` sounds like a good option.

**Comment by nad** [comment]
I don't think we should implement a more general variant unless there is some use for the extra generality (in the near future). Do you have some use case in mind, other than `--safe-except-for=sized-types`?

People have looked into fixing some of the problems with sized types during the current Agda meeting. If the consistency issues are fixed, then I suggest that we close this issue without taking any action.

**Comment by turion** [comment]
> I don't think we should implement a more general variant unless there is some use for the extra generality (in the near future). Do you have some use case in mind, other than --safe-except-for=sized-types?

For any other flag that is unsafe, and it seems like there are a lot. Maybe `--safe` should simply be implemented so it can be overridden with unsafe flags.

> People have looked into fixing some of the problems with sized types during the current Agda meeting. If the consistency issues are fixed, then I suggest that we close this issue without taking any action.

Who else is working on fixing other issues currently? I'm not aware of any other action currently. My impression is that the consistency issues are far from being fixed.

**Comment by nad** [comment]
> For any other flag that is unsafe, and it seems like there are a lot.

Again, do you have a use case for this? Can you give a concrete example of an unsafe flag for which you would want to do this?

> Maybe `--safe` should simply be implemented so it can be overridden with unsafe flags.

I don't think there should be a difference between `--safe --no-termination-check` and `--no-termination-check --safe`. Furthermore, if I run `agda --safe M.agda`, then Agda should complain if `M.agda` starts with `{-# OPTIONS --no-termination-check #-}`.

> Who else is working on fixing other issues currently?

When I wrote that I did not know who was behind the nick "turion".

**Comment by turion** [comment]
> > For any other flag that is unsafe, and it seems like there are a lot.
> 
> Again, do you have a use case for this? Can you give a concrete example of an unsafe flag for which you would want to do this?

Not a concrete use case I'm working on right now. My suggestion was more to have a coherent, general mechanism instead of patching one particular use case. Usually, people have lots of use cases, but only rarely speak up in the form of a Github issue to voice their needs.

> 
> > Maybe `--safe` should simply be implemented so it can be overridden with unsafe flags.
> 
> I don't think there should be a difference between `--safe --no-termination-check` and `--no-termination-check --safe`.

I agree. How about this: if there is the `--safe` flag and an unsafe flag _on the command line_, then agda should not complain if `{-# OPTIONS --no-termination-check #-}` is used in the code in some places.

> Furthermore, if I run `agda --safe M.agda`, then Agda should complain if `M.agda` starts with `{-# OPTIONS --no-termination-check #-}`.

True, that's the big advantage behind `--safe`.

> > Who else is working on fixing other issues currently?
> 
> When I wrote that I did not know who was behind the nick "turion".

Sorry, I didn't make that clear. Since I'm new to the implementation side of Agda, I'm not expecting to be solving any consistency issues soon yet. But if anyone else is, that's great news :)

**Comment by jespercockx** [comment]
Perhaps what we need is not a flag `--safe-except-for` but rather a flag `--report-unsafe` that prints out a list of all unsafe features that were used at the very end of the log in BIG FAT LETTERS. That way you can make an informed decision on whether each of the unsafe features used is acceptable or not. For some unsafe features such as `postulate` it could even print out a list of all postulates and their location.

**Comment by nad** [comment]
Perhaps `--report-unsafe` would be useful, but it does not cover my use case for `--safe-except-for-sized-types`: I invoke Agda from another program, and I want Agda to exit with a non-zero exit code when unsafe features other than sized types are used.

**Comment by jespercockx** [comment]
Maybe the output of `--report-unsafe` could be in some standard format so it would be easy to automatically analyze it and throw an error when there's any unexpected unsafe features.

**Comment by jespercockx** [comment]
How about rather than having various `--safe-except-for-...` flags, we would have `--safe` simply be an alias to a bunch of other flags that can be enabled and disabled individually, like `--no-postulates`, `--no-terminating-pragmas`, ... (plus a number of flags for turning warnings into errors)? That way you could "build your own --safe flag" by enabling the specific checks that you want.

**Comment by nad** [comment]
In that case, if some new unsafe feature is added to Agda (and turned on by default), then you would have to switch to another set of flags. Perhaps it would be better to make Agda safe by default, and use explicit flags (incompatible with `--safe`) to enable unsafe features.

**Comment by jespercockx** [comment]
But disabling all unsafe features by default wouldn't really serve the same purpose: as I understand the reason to use `--safe` is that it ensures *no unsafe features are enabled in this file or any of its imports*. While if we would (say) have a flag to enable postulates which is disabled by default, there is nothing that prevents one of the imports from enabling it. Or do you propose that all the flags to enable unsafe features should be infective? Either way, requiring flags for unsafe features would be a huge breaking change.

**Comment by jespercockx** [comment]
I think a reasonable compromise that does not break backwards compatibility (and is not too far from what we're already doing) would be the following policy:
- *Old* unsafe features such as `postulate`s, `{-# TERMINATING #-}`, `{-# NO_POSITIVITY_CHECK #-}`, `--sized-types` etc are *allowed by default* but get a flag for turning them off.
- *New* unsafe features should have flags for turning them on/off, and should be *off by default*
- `--safe` becomes an alias for a a set of flags including (1) turning off all the old unsafe features, and (2) explicitly turning off all new unsafe features.

That way you can "build your own --safe flag" by picking which unsafe features to enable/disable, and still be sure that future updates of Agda do not allow new unsafe features from being silently enabled.

**Comment by nad** [comment]
Sounds good to me.

**Comment by jespercockx** [comment]
We have to add the following flags:
- `--no-compile-pragmas`: turn off COMPILE pragmas (in the scope checker)
- `--safe-termination-check`: don't allow turning off termination (in the scope checker)
- `--safe-positivity-check`: don't allow turning off positivity checking (in the scope checker)
- `--no-postulates`: disallow postulates (in the scope checker & reflection API)
- `--no-erased-equality-without-K`: don't allow `primEraseEquality` when --without-K

EDIT: Here are some more flags we'd need to add:
- `--no-polarity-pragmas`
- `--no-injective-pragmas`
- `--no-eta-pragmas`
- `--no-cubical-with-K`: this flag wouldn't have any real effect but would just prevent `--cubical` and `--with-K` from being used together in the same file.
- `--no-unsolved-metas`
- `--safe-coverage-check`: this flag would disable `--allow-incomplete-matches`, `{-# NO_COVERAGE_CHECK #-}` pragmas, and cause missing definitions to raise an error.
- `--safe-universe-check`: forbid the use of `--type-in-type`, `--omega-in-omega`, and `{-# NO_UNIVERSE_CHECK #-}` pragmas

**Comment by jespercockx** [comment]
I've updated my comment above with some more flags that we'd need to add to reach the goal of "having a modular --safe flag". As you can see it's starting to become quite a long list. Of course we can discuss the naming of these flags, as well as whether some of them should be merged or split up further. However it's clear to me now that this is not the "quick fix" I thought it was at first.

More generally, it would make sense to do an overhaul of the `--safe` flag where we keep track of the set of unsafe features that are used in each file and its imports. That way we could introduce new flags such as `--no-postulates` without making them coinfective (since we now have an easy way to check if a file or any of its imports has postulates). This would even allow us to remove the current coinfective status of `--safe` itself. This would in particular allow importing of modules that are safe but not explicitly marked as `--safe` from one that does have `--safe`. It would also greatly reduce how often Agda needs to recheck dependencies when an option is changed.

I'd be willing to do this overhaul but it doesn't feel like a good idea to rush it before the release of 2.6.2. However, since this overhaul would have a big impact on the semantics on all the flags that would be introduced by the granularization of `--safe`, I'm hesitant to introduce these new flags at the moment. Perhaps we can implement `--safe-except-for-sized-types` as a stopgap measure after all, and keep the bigger overhaul of `--safe` for 2.6.3.

**Comment by jespercockx** [comment]
This is blocked by #5265

**Comment by nad** [comment]
We should perhaps mention that if one simply removes `--safe` from `{-# OPTIONS --safe --sized-types #-}`, then one also turns on `--guardedness` (in the absence of options given on the command-line or in a `.agda-lib` file). What is currently the best approximation of `--safe-except-for-sized-types --sized-types`? Is it `--sized-types --no-guardedness`?

**Comment by jespercockx** [comment]
Perhaps we should consider turning off `--guardedness` by default when enabling `--sized-types`.

**Comment by nad** [comment]
Yes, that makes sense. Let's do this before we release Agda 2.6.2.

**Comment by nad** [comment]
An alternative would be to make `--no-guardedness` the default even in the absence of `--sized-types`. This would break some code, but it might be less confusing: I prefer "`--guardedness` is off by default" to "`--guardedness` is on by default, unless `--safe` or `--sized-types` is used".

**Comment by jespercockx** [comment]
In theory I agree, but I'm a bit scared of the amount of code that would break. Most people see guardedness as the default option, which will be even more so now that sized types are no longer considered safe.

**Comment by jespercockx** [comment]
I realized making `--sized-types` turn off `--guardedness` only makes sense if `--sized-types` is no longer default. So it seems we have three options:

1. Make `--no-sized-types` the default
2. Make `--no-guardedness` the default
3. Keep the current situation (both `--sized-types` and `--guardedness` are on by default)

My vote would go to option 1 as there are no known soundness issues with `--guardedness`, while there are for `--sized-types`. Also there is no special syntax associated with `--guardedness`, while `--sized-types` is only useful if you actually import and use the primitives. Also, with the new possibility of adding flags to agda-lib files, the burden of adding explicit `--sized-types` flags shouldn't be too big.

Since this seems like an important decision, I'd like other people's opinions as well, in particular @andreasabel and @UlfNorell .

**Comment by nad** [comment]
There are perhaps no known soundness issues with `--guardedness`, but it is not backed by solid theory (see #5273). I think I would prefer to make both `--no-sized-types` and `--no-guardedness` the default.

**Comment by jespercockx** [comment]
Indeed, that is a fourth option. Actually I have no idea how much code out there relies on guardedness; I was under the impression that it is quite a lot but maybe that's not the case? I'll have to do the experiment I guess.

**Comment by jespercockx** [comment]
I was trying out the approach where `--sized-types` would turn off `--guardedness`, but this makes it impossible to import a module with no flags from a module that uses `--sized-types` (because the module with no flags has an implicit `--guardedness`, while the module with `--sized types` has an implicit `--no-guardedness`). This in turn requires a lot of modules in the standard library to get the additional flag `--no-guardedness`. So perhaps the right solution is indeed to make neither of them enabled by default.

**Comment by jespercockx** [comment]
Here is the impact on the standard library of disabling both `--sized-types` and `--guardedness` by default: https://github.com/jespercockx/agda-stdlib/commit/2c8a40ef524e0d809396d33cff52863580ad75b5
I'm preparing a PR so we can assess the impact on the Agda codebase itself as well.

**Comment by jespercockx** [comment]
See https://github.com/agda/agda/pull/5386. Overall the impact of disabling --sized-types by default seems to be much larger than disabling --guardedness by default, at least on the test suite and the standard library. So maybe just disabling --guardedness by default always is the right way to go.

**Comment by nad** [comment]
You added `--no-guardedness` to some standard library modules with the `--safe` flag. That should not be necessary.

The changes to the standard library seem to be fairly small, and I think we can ignore the changes to the test suite.

**Comment by jespercockx** [comment]
I have done everything on this issue that I planned to do before the release. I'm still working on a bigger refactoring that will modularize the --safe flag (and will have some other benefits as well, such as being able to import modules that do not use --safe as long as they do not use unsafe features), but that will have to wait until 2.6.3.
